ocaml/src/blob_access.ml

```diff
@@ -0,0 +1,384 @@
+
+
+  method send_blob_data_to fnr size slices nfd =
+    Lwt_extra2.with_fd
+      (if config.write_blobs
+       then self # _get_file_path fnr
+       else "/dev/zero")
+      ~flags:Lwt_unix.([O_RDONLY;])
+      ~perm:0600
+      (fun blob_fd ->
+        let blob_ufd = Lwt_unix.unix_file_descr blob_fd in
+        let () =
+          if config.use_fadvise
+          then
+            begin
+              Posix.posix_fadvise blob_ufd 0 size Posix.POSIX_FADV_RANDOM;
+              List.iter
+                (fun (offset, length) ->
+                  Posix.posix_fadvise
+                    blob_ufd
+                    offset length
+                    Posix.POSIX_FADV_WILLNEED)
+                slices
+            end
+        in
+        Lwt_list.iter_s
+          (fun (offset, length) ->
+            Net_fd.sendfile_all
+              ~fd_in:blob_fd ~offset
+              ~fd_out:nfd
+              length
+            >>= fun () ->
+            let () =
+              if config.use_fadvise
+              then
+                Posix.posix_fadvise
+                  blob_ufd
+                  0 size
+                  Posix.POSIX_FADV_DONTNEED
+            in
+            Lwt.return_unit)
+          slices
+      )
+  method push_blob_data fnr size slices f =
+    Lwt_extra2.with_fd
+      (if config.write_blobs
+       then self # _get_file_path fnr
+       else "/dev/zero")
+      ~flags:Lwt_unix.([O_RDONLY;])
+      ~perm:0600
+      (fun blob_fd ->
+        let blob_ufd = Lwt_unix.unix_file_descr blob_fd in
+        let () =
+          if config.use_fadvise
+          then
+            begin
+              Posix.posix_fadvise blob_ufd 0 size Posix.POSIX_FADV_RANDOM;
+              List.iter
+                (fun (offset, length) ->
+                  Posix.posix_fadvise
+                    blob_ufd
+                    offset length
+                    Posix.POSIX_FADV_WILLNEED)
+                slices
+            end
+        in
+        Lwt_list.iter_s
+          (fun (offset, length) ->
+            let buffer = Lwt_bytes.create length in
+            (* fill it *)
+            f (offset,length) buffer 0
+            >>= fun () ->
+            let () =
+              if config.use_fadvise
+              then
+                Posix.posix_fadvise
+                  blob_ufd
+                  0 size
+                  Posix.POSIX_FADV_DONTNEED
+            in
+            Lwt.return_unit)
+          slices
+      )
+
+
+  method private _with_blob_fd fnr f =
+    Lwt_extra2.with_fd
+      (if config.write_blobs
+       then self # _get_file_path fnr
+       else "/dev/zero")
+      ~flags:Lwt_unix.([O_RDONLY;])
+      ~perm:0600
+      f
+
+
+  method get_blob fnr size =
+    Lwt_log.debug_f "getting blob %Li with size %i" fnr size >>= fun () ->
+    let bs = Bytes.create size in
+    self # _with_blob_fd
+      fnr
+      (fun fd ->
+         Lwt_extra2.read_all fd bs 0 size >>= fun got ->
+         assert (got = size);
```

Comment:
There's a `Lwt_extra2.read_all_exact` method which combines the read and the assert